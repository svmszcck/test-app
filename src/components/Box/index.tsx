import React from "react";
import { View } from "react-native";

const Box = () => {
  return <View></View>;
};

export default Box;
