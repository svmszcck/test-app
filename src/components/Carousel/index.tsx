import React from "react";
import {
  View,
  FlatList,
  StyleSheet,
  ActivityIndicator,
  GestureResponderEvent,
} from "react-native";
import { Text, Tile } from "react-native-elements";

import { IMAGE_URL } from "app_constants/api";
import Device from "app_constants/layout";
import { IMAGE_MEDIUM, POSTER_TEXT_LIMIT } from "app_constants/ui";
import Routes from "app_constants/routes";
import { trimText } from "utils/ui";
import { useColor } from "hooks";
import { Movie } from "types";

const posterWidth = Device.window.width / 2.5;
const posterHeight = posterWidth / 1.5;

const Carousel = ({
  elements = [],
  navigation,
  isLoading = false,
  loadMore,
}: CarouselType) => {
  const colors = useColor();
  return (
    <FlatList
      keyExtractor={(item) => item.id}
      data={elements}
      horizontal
      showsHorizontalScrollIndicator={false}
      renderItem={({
        item: { id, title, poster_path: posterPath, vote_average: voteAverage },
      }) => (
        <View style={styles.poster}>
          <Tile
            imageSrc={{
              uri: `${IMAGE_URL(IMAGE_MEDIUM)}/${posterPath}`,
            }}
            featured
            activeOpacity={0.8}
            height={posterHeight}
            width={posterWidth}
            onPress={() => navigation.navigate(Routes.MOVIE_DETAILS, { id })}
            imageContainerStyle={styles.posterImage}
            containerStyle={styles.posterContainer}
          />
          <Text style={[styles.posterTitle, { color: colors.text }]}>
            {title.length >= POSTER_TEXT_LIMIT
              ? trimText(title, POSTER_TEXT_LIMIT)
              : title}
          </Text>
        </View>
      )}
      ListFooterComponent={
        <View style={styles.posterLoadMore}>
          {isLoading ? (
            <ActivityIndicator animating size="large" color={colors.text} />
          ) : (
            <Text
              style={[styles.posterLoadMoreTxt, { color: colors.textBold }]}
              onPress={loadMore}
            >
              Load More
            </Text>
          )}
        </View>
      }
    />
  );
};

const styles = StyleSheet.create({
  poster: {
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center",
    marginRight: 20,
  },
  posterContainer: {
    justifyContent: "center",
    alignItems: "center",
    borderRadius: 10,
  },
  posterImage: {
    borderRadius: 10,
  },
  posterTitle: {
    marginTop: 5,
    fontSize: 15,
  },
  posterLoadMore: {
    justifyContent: "center",
    alignItems: "center",
    flex: 1,
    minWidth: 150,
    marginRight: 20,
  },
  posterLoadMoreTxt: {
    fontWeight: "bold",
    fontSize: 18,
  },
});

type CarouselType = {
  elements: Array<Movie>;
  navigation: any;
  isLoading: boolean;
  loadMore: (event: GestureResponderEvent) => void;
};

export default Carousel;
