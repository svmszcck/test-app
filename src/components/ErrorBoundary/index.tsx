import React, { Component, ErrorInfo } from "react";
import { View, StyleSheet } from "react-native";
import * as Updates from "expo-updates";
import { Text, Button } from "react-native-elements";

import { Theme } from "app_constants/colors";

class ErrorBoundary extends Component {
  state = {
    errorMsg: "",
  };

  static getDerivedStateFromError(error: Error) {
    return { errorMsg: error.toString() };
  }

  componentDidCatch(error: Error, info: ErrorInfo) {
    console.error("Uncaught error: ", error, info);
  }

  restartApp() {
    Updates.reloadAsync();
  }

  render() {
    if (this.state.errorMsg)
      return (
        <View style={styles.container}>
          <Text style={[styles.warning, { color: Theme.textBold }]}>
            Something went wrong :(
          </Text>
          <Button title="Restart App" onPress={this.restartApp} />
        </View>
      );
    return this.props.children;
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  warning: {
    fontSize: 20,
    marginBottom: 40,
  },
});

export default ErrorBoundary;
