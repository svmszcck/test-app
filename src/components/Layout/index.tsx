import React, { useMemo } from "react";
import {
  View,
  ActivityIndicator,
  StyleSheet,
  TouchableOpacity,
  useColorScheme,
} from "react-native";
import { Text } from "react-native-elements";
import { Ionicons } from "@expo/vector-icons";

import Colors from "constants/colors";

const Layout = ({
  children,
  isLoading = false,
  hasMenu = false,
  rightIcon,
  rightAction,
  navigation,
}: LayoutProps) => {
  const colorScheme = useColorScheme();
  const colors = useMemo(() => Colors[colorScheme], []);
  return (
    <View style={[styles.container, { backgroundColor: colors.primary }]}>
      {isLoading ? (
        <ActivityIndicator
          animating
          size="large"
          color={colors.text}
          style={styles.loading}
        />
      ) : (
        <>
          {hasMenu && (
            <View style={styles.header}>
              <TouchableOpacity
                style={{ width: 60, height: 60 }}
                onPress={() => navigation.goBack()}
              >
                <Ionicons size={30} color={colors.text} name="ios-arrow-back" />
              </TouchableOpacity>
              <TouchableOpacity style={styles.favorite} onPress={rightAction}>
                <Ionicons size={30} name={rightIcon} />
              </TouchableOpacity>
            </View>
          )}
          {children}
        </>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    paddingBottom: 0,
  },
  loading: {
    position: "absolute",
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  header: {
    flexDirection: "row",
    marginBottom: 20,
  },
  favorite: {
    marginLeft: "auto",
    width: 30,
  },
});

type LayoutProps = {
  children: React.ReactNode;
  isLoading: boolean;
};

export default Layout;
