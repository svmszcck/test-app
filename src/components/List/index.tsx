import React from "react";
import {
  View,
  ScrollView,
  ActivityIndicator,
  StyleSheet,
  FlatList,
} from "react-native";
import { Avatar, Button, Text, ListItem } from "react-native-elements";
import { isEmpty } from "lodash";
import { Ionicons } from "@expo/vector-icons";

import { IMAGE_URL } from "app_constants/api";
import Routes from "app_constants/routes";
import { IMAGE_SMALL } from "app_constants/ui";
import { DELAY_TIME, LOAD_MORE } from "app_constants/general";
import { delayTask } from "utils/ui";
import { Movie } from "types";
import { useColor } from "hooks";

import placeholder from "assets/images/movie.jpg";

const List = ({
  navigation,
  elements,
  hasAction = false,
  icon,
  rightAction,
  hasLoadMore = false,
  isLoading = false,
  loadMore,
  warning,
}: ListProps) => {
  const colors = useColor();

  return (
    <View>
      {!isLoading && isEmpty(elements) && warning && (
        <Text style={[styles.warning, { color: colors.textBold }]}>
          {warning}
        </Text>
      )}
      <ScrollView
        showsVerticalScrollIndicator={false}
        contentContainerStyle={styles.container}
      >
        <FlatList
          keyExtractor={(item) => item.id.toString()}
          data={elements}
          showsVerticalScrollIndicator={false}
          renderItem={({
            item: {
              id,
              title,
              poster_path: posterPath,
              vote_average: voteAverage,
            },
          }) => (
            <ListItem
              style={styles.item}
              onPress={() => navigation.navigate(Routes.MOVIE_DETAILS, { id })}
            >
              <Avatar
                size="large"
                source={
                  posterPath
                    ? {
                        uri: `${IMAGE_URL(IMAGE_SMALL)}/${posterPath}`,
                      }
                    : placeholder
                }
              />
              <ListItem.Content>
                <ListItem.Title>{title}</ListItem.Title>
                <ListItem.Subtitle style={{ color: colors.text }}>
                  Score: {voteAverage ? voteAverage : "-"}
                </ListItem.Subtitle>
              </ListItem.Content>
              {hasAction ? (
                <Ionicons
                  onPress={() => rightAction(id)}
                  size={30}
                  color={colors.text}
                  name={icon}
                />
              ) : (
                <ListItem.Chevron />
              )}
            </ListItem>
          )}
        />
        <View style={styles.bottomSection}>
          {hasLoadMore && !isLoading && !isEmpty(elements) && (
            <Button
              buttonStyle={[
                styles.loadMore,
                { backgroundColor: colors.secondary },
              ]}
              title={LOAD_MORE}
              onPress={() => delayTask(loadMore, DELAY_TIME)}
            />
          )}
          {isLoading && (
            <ActivityIndicator
              animating
              size="large"
              color={colors.text}
              style={styles.loading}
            />
          )}
        </View>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingBottom: 40,
  },
  warning: {
    marginTop: 30,
    marginLeft: "auto",
    marginRight: "auto",
    fontWeight: "bold",
  },
  bottomSection: {
    height: 80,
    alignItems: "center",
    justifyContent: "center",
  },
  loadMore: {
    margin: 20,
    backgroundColor: "white",
  },
  item: {
    marginBottom: 12,
  },
  loading: {
    marginTop: 20,
  },
});

type ListProps = {
  navigation: any;
  elements: Array<Movie>;
  hasAction: boolean;
  icon: string;
  rightAction: Function;
  hasLoadMore: boolean;
  isLoading: boolean;
  loadMore: Function;
  warning: string;
};

export default List;
