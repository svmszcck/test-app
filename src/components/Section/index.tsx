import React, { useMemo, ReactNode } from "react";
import { View, StyleSheet, useColorScheme } from "react-native";
import { Text } from "react-native-elements";

import Colors from "app_constants/colors";

const Section = ({ text, children, style }: SectionProps) => {
  const colorScheme = useColorScheme();
  const colors = useMemo(() => Colors[colorScheme], []);
  return (
    <View style={styles.container}>
      <Text style={[styles.label, { color: colors.textBold, ...style }]}>
        {text}
      </Text>
      {children}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginBottom: 20,
  },
  label: {
    marginBottom: 10,
    fontSize: 25,
  },
});

type SectionProps = {
  text: string;
  children: ReactNode;
  style: Object;
};

export default Section;
