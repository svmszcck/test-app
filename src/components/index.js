export { default as Layout } from "./Layout";
export { default as Section } from "./Section";
export { default as List } from "./List";
export { default as Carousel } from "./Carousel";
export { default as ErrorBoundary } from "./ErrorBoundary";
