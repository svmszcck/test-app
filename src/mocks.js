import React from "react";
import configureMockStore from "redux-mock-store";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";
import thunk from "redux-thunk";

const mockStore = configureMockStore([thunk]);
const Stack = createStackNavigator();

export const store = mockStore({
  user: {
    name: "Onur",
    avatar: null,
    favorites: [],
  },
  posts: {
    movie: {
      id: 12,
      title: "Rwer",
      poster_path: "ekjrhtkjert",
      vote_average: 5.6,
    },
  },
});

export const MockedNavigator = ({ component, params = {} }) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          name="MockedScreen"
          component={component}
          initialParams={params}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export const navigation = {
  navigate: () => {},
};
