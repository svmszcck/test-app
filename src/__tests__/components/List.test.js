import "react-native";
import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

import { List } from "components";
import { store } from "mocks";

jest.useFakeTimers();

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <List />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
