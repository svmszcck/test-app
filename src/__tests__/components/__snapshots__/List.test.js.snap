// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "paddingBottom": 40,
      }
    }
    showsVerticalScrollIndicator={false}
  >
    <View>
      <RCTScrollView
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
